### ISSUE TITLE
Rule proposal: lint readonly not-assigned properties

### ISSUE THREAD

**Issue by jendrikw** [typescript-eslint/typescript-eslint#12315]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [x] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

I have a readonly class property that I forgot to assign in the constructor. Even with `strictPropertyInitialization`, typescript does not warn when the type of the property allows undefined.

The result is that the property can only ever be undefined, which is probably not what the author intended (and in such a case should be typed as just `undefined`.)

### Fail Cases

```TypeScript
class A {
  readonly x: string | undefined;
  constructor(x: string | undefined) {
    // complex logic
    if (x == "foo") {
      x = "bar"
    }
    // forget to intialize x
  }
}
```

### Pass Cases

```TypeScript
class A {
  readonly x: string | undefined;
  constructor(x: string | undefined) {
    // complex logic
    if (x == "foo") {
      x = "bar"
    }
    this.x = x;
  }
}
```

**Comment by bradzacher** [comment]
I don't think that this is generally possible to lint for. Because it's impossible for us to know if it was intentional or not that the property was not assigned.

In general as well this is such a rare case that I don't think there is sufficient value in linting for it.

If anything I would suggest an upstream issue with TS if you think this is a gap in the existing compiler option's behaviour.
